pkg/clients/clients.go

```diff
@@ -1,57 +1,161 @@
 	"k8s.io/apimachinery/pkg/runtime/schema"
 	"k8s.io/client-go/discovery"
 	"k8s.io/client-go/kubernetes"
-	"k8s.io/client-go/rest"
+	restclient "k8s.io/client-go/rest"
 	"k8s.io/client-go/tools/record"
+	controllerruntime "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
 	"sigs.k8s.io/controller-runtime/pkg/log/zap"
 )
 
+//go:generate mockgen -source=clients.go -package=clients -destination=mock_clients_api.go
+
+const (
+	clusterVersionName = "version"
+)
+
 var (
-	log        = zap.New(zap.UseDevMode(true)).WithName(color.Print("clients", color.Brown))
-	Interface  *ClientsInterface
-	RestConfig *rest.Config
-	Namespace  string
-	config     genericclioptions.ConfigFlags
+	log       = zap.New(zap.UseDevMode(true)).WithName(color.Print("clients", color.Brown))
+	Interface ClientsInterface
+	// TODO need to remove this global variable
+	Namespace string
 )
 
-type ClientsInterface struct {
-	client.Client
-	kubernetes.Clientset
-	clientconfigv1.ConfigV1Client
-	record.EventRecorder
-	authn.Keychain
-	discovery.CachedDiscoveryInterface
+type ClientsInterface interface {
+	Update(ctx context.Context, obj client.Object) error
+	Get(ctx context.Context, key client.ObjectKey, obj client.Object) error
+	Delete(ctx context.Context, obj client.Object) error
+	List(ctx context.Context, obj client.ObjectList, opts ...client.ListOption) error
+	Create(ctx context.Context, obj client.Object) error
+	GetPodLogs(namespace, podName string, podLogOpts *v1.PodLogOptions) *restclient.Request
+	GetNamespace(ctx context.Context, name string, opts metav1.GetOptions) (*v1.Namespace, error)
+	GetSecret(ctx context.Context, namespace, name string, opts metav1.GetOptions) (*v1.Secret, error)
+	ClusterOperatorCreate(ctx context.Context, clusterOperator *configv1.ClusterOperator, opts metav1.CreateOptions) (*configv1.ClusterOperator, error)
+	ClusterOperatorGet(ctx context.Context, name string, opts metav1.GetOptions) (*configv1.ClusterOperator, error)
+	ClusterOperatorUpdateStatus(ctx context.Context, clusterOperator *configv1.ClusterOperator, opts metav1.UpdateOptions) (*configv1.ClusterOperator, error)
+	ClusterVersionGet(ctx context.Context, opts metav1.GetOptions) (result *configv1.ClusterVersion, err error)
+	Invalidate()
+	ServerGroups() (*metav1.APIGroupList, error)
+	StatusUpdate(ctx context.Context, obj client.Object) error
+	CreateOrUpdate(ctx context.Context, obj client.Object, fn controllerutil.MutateFn) (controllerutil.OperationResult, error)
+	HasResource(resource schema.GroupVersionResource) (bool, error)
+	GetPlatform() (string, error)
 }
 
-// GetKubeClientSet returns a native non-caching client for advanced CRUD operations
-func GetKubeClientSet() (*kubernetes.Clientset, error) {
-	return kubernetes.NewForConfig(RestConfig)
+type k8sClients struct {
+	runtimeClient   client.Client
+	clientset       kubernetes.Clientset
```

Comment:
Could this be a `kubernetes.Interface`?

Reply:
Clientset is a struct that implements kuberentes.Interface

Reply:
@yevgeny-shnaidman better have the interface there then, as we can mock it if we need to